<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Introduction to epivizr: interactive visualization for epigenomics</title>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: rgb(88, 72, 246)
   }

   pre .number {
     color: rgb(0, 0, 205);
   }

   pre .comment {
     color: rgb(76, 136, 107);
   }

   pre .keyword {
     color: rgb(0, 0, 255);
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: rgb(3, 106, 7);
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>

<!-- MathJax scripts -->
<script type="text/javascript" src="https://c328740.ssl.cf1.rackcdn.com/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>


<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}
</style>



</head>

<body>
<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{Introduction to epivizr}
-->

<h1>Introduction to epivizr: interactive visualization for epigenomics</h1>

<p><a href="http://epiviz.cbcb.umd.edu"><code>Epiviz</code></a> is an interactive visualization tool for epigenomic data. It supports genome navigation like other genome browsers, but allows multiple visualizations of data within genomic regions using scatterplots, heatmaps and other user-supplied visualizations. It also includes data from the <a href="http://barcode.luhs.org/">Gene Expression Barcode project</a> for transcriptome visualization. It also has a flexible plugin framework so users can add <a href="http://d3js.org/">d3</a> visualizations. You can find more information about Epiviz at <a href="http://epiviz.cbcb.umd.edu/help">http://epiviz.cbcb.umd.edu/help</a> and see a video tour <a href="http://youtu.be/099c4wUxozA">here</a>.</p>

<iframe width="480" height="360" src="http://www.youtube.com/embed/099c4wUxozA" frameborder="1" allowfullscreen></iframe>

<p>The <code>epivizr</code> package implements two-way communication between the <code>R/Bioconductor</code> computational genomics environment and <code>Epiviz</code>. Objects in an <code>R</code> session can be displayed as tracks or plots on Epiviz. Epivizr uses Websockets for communication between the browser Javascript client and the R environment using the same technology underlying the popular <a href="http://www.rstudio.com/shiny/">Shiny</a> system for authoring interactive web-based reports in R.</p>

<h3>Preliminaries: the data</h3>

<p>In this vignette we will look at colon cancer methylation data from the TCGA project and expression data from the gene expression barcode project. We include methylation data from the Illumina 450kHumanMethylation beadarray processed with the <code>minfi</code> package in <code>epivizr</code>. We get the expression data from the <code>antiProfilesData</code> bioconductor package.</p>

<pre><code class="r">require(epivizr)
require(antiProfilesData)
</code></pre>

<pre><code class="r">data(tcga_colon_example)
data(apColonData)
</code></pre>

<p>The <code>colon_blocks</code> object is a <code>GRanges</code> object containing regions of hypo or hyper methylation in colon cancer identified using the <code>blockFinder</code> function in the <code>minfi</code> package. </p>

<pre><code class="r">show(colon_blocks)
</code></pre>

<pre><code>## GRanges with 2887 ranges and 11 metadata columns:
##          seqnames                 ranges strand   |               value
##             &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt;   |           &lt;numeric&gt;
##      [1]    chr15 [ 24778439,  28216142]      *   |  -0.120897240685778
##      [2]    chr11 [  4407026,   6435089]      *   |  -0.142954896408814
##      [3]     chr1 [151988500, 153479701]      *   |  -0.163755989232385
##      [4]     chr2 [217468708, 219096237]      *   |  -0.142798859426357
##      [5]    chr11 [131239366, 133716186]      *   |  -0.135137323912239
##      ...      ...                    ...    ... ...                 ...
##   [2883]     chr3 [189314757, 189314757]      *   | -0.0661840993557387
##   [2884]    chr10 [ 51525504,  51525504]      *   | -0.0661709050441534
##   [2885]     chr5 [ 81936043,  81936043]      *   |  0.0660822894241233
##   [2886]    chr22 [ 25170766,  25170766]      *   | -0.0657897823458402
##   [2887]     chr7 [156032994, 156032994]      *   | -0.0657554173060194
##                        area   cluster indexStart  indexEnd         L
##                   &lt;numeric&gt; &lt;numeric&gt;  &lt;integer&gt; &lt;integer&gt; &lt;numeric&gt;
##      [1]   32.1586660224169       614     167155    167531       266
##      [2]   30.3064380386685       495     130755    131000       212
##      [3]   25.3821783310197        32      13751     13939       155
##      [4]   25.2753981184652        93      36420     36636       177
##      [5]   23.9193063324663       520     141959    142173       177
##      ...                ...       ...        ...       ...       ...
##   [2883] 0.0661840993557387       145      51658     51658         1
##   [2884] 0.0661709050441534       474     122060    122060         1
##   [2885] 0.0660822894241233       263      67322     67322         1
##   [2886] 0.0657897823458402       839     216913    216913         1
##   [2887] 0.0657554173060194       356     101022    101022         1
##           clusterL           p.value      fwer       p.valueArea  fwerArea
##          &lt;integer&gt;         &lt;numeric&gt; &lt;numeric&gt;         &lt;numeric&gt; &lt;numeric&gt;
##      [1]       664                 0         0                 0         0
##      [2]      1629                 0         0                 0         0
##      [3]      1651                 0         0                 0         0
##      [4]       527                 0         0                 0         0
##      [5]      1759                 0         0                 0         0
##      ...       ...               ...       ...               ...       ...
##   [2883]      1125 0.976855856124549         1 0.982015568612246         1
##   [2884]       269 0.977467267141877         1 0.982462942527364         1
##   [2885]       374 0.982462942527364         1 0.986131408631334         1
##   [2886]        85 0.997807867815921         1 0.998046467237317         1
##   [2887]       468 0.999746488114766         1 0.999761400578604         1
##   ---
##   seqlengths:
##     chr1 chr10 chr11 chr12 chr13 chr14 ...  chr6  chr7  chr8  chr9  chrX
##       NA    NA    NA    NA    NA    NA ...    NA    NA    NA    NA    NA
</code></pre>

<p>The columns <code>value</code> and <code>p.value</code> can be used to determine which of these regions, or blocks, are interesting by looking at a volcano plot for instance.</p>

<pre><code class="r">plot(colon_blocks$value, -log10(colon_blocks$p.value), main = &quot;Volcano plot&quot;, 
    xlab = &quot;Avg. methylation difference&quot;, ylab = &quot;-log10 p-value&quot;, xlim = c(-0.5, 
        0.5))
</code></pre>

<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAADAFBMVEUAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACzMPSIAAABAHRSTlMAAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+f4CBgoOEhYaHiImKi4yNjo+QkZKTlJWWl5iZmpucnZ6foKGio6SlpqeoqaqrrK2ur7CxsrO0tba3uLm6u7y9vr/AwcLDxMXGx8jJysvMzc7P0NHS09TV1tfY2drb3N3e3+Dh4uPk5ebn6Onq6+zt7u/w8fLz9PX29/j5+vv8/f7/qVjM+gAAAAlwSFlzAAALEgAACxIB0t1+/AAAHBFJREFUeJztnQl8DGcfx5/NRVCCoOpoq3UfbalW+1J1lGqltBTVlKK0VC9VenhLW3U1jmgllKqjSYkgCYlI5EBEqAZxFImQO7JJNtn7mJnnndlzdnd295nZSez2ne+HNfs8z/7nPz/PPMc8xwAgICAgICAgICAgICAgICAgICAgICAgICAgICDQACyH+8jPQBXsaQ5qCiFbK9Y/6dXLhwfPPIT+sMYXgJdhniXIbYEgbMqDZx6CKB8+C8BGuNwSJAhkxRpKm+uwL6nVwkL89ociw9XOvI5VhjUBHaOq8NqEnlTYpDtYTh9LKgoydHaNZKWP4SemGEjS6z5fFY8MhmdBF3iDvOI5UJWsgrP1VzsVqhKL4c8gDVbGX4bXqTDlBSnMtqSiIEMVZ3VwoUEgU8wNCG91u89XxSOiYrzte3AleXQNTgPTyMKIutrz8C3QUSXx+TO2D2gHYQAZNgo8CTFLKgoydCaYBS8bBDLH/LtuMbAJTj0AnwTAj4DNwQMQ96WuVgPb6iN9J0dk1JIXTIaJQCAZYU5FxZKhLUBL408sMf8ygV6AeyUF5B3mQ11gC0gYBWqjj9wJ0z4ebRDIUBSbU1GxZAD1xSCQJeZfJpDvPR1cSx1cgVPNt9hF8t7prK7xUcP2YChdIHMqCjJghuUWM8dAGHhfr4hvtkE4mPqXLGSPqUhhqKt9B2qTSuA6cAfeSJSS+cQikCkVhbGQnmsTo4LRHe/rFfHMGFikr7VFC27rCuaZdLihK1/pB4bm4Xmv1MGpFoFMqSjIgDmSum9FNjFfK9SP3cfr8SA4NCn/vxAEckGTu3fvtwsCAgICAgICAgICAgICAgICAgICAgINx1YA+uRq07vcbz88FghA9orAOUm24W3e/FcwyZcHgbBmQFRhDpiYquf27Xn/Bgq7ui1QB9G1zqB5mW34L3+6a9kjiHBboJMl8roNIDPcNhxFINHjA5owxwQP6jLms/kzZ38e/kqzSMm9y5U4tmjynKGDOyP71WFga+S0DvDv19PHfYEAaNqzN1jgbxuKIFDLtK0rc/7DFPNd4j4NpoMEhBgBTRBSvDxphwjNqQ1xy098ipbUEQNy1oSf2sWDQMwgCBTxHAAtMhhKwRHrg66dK6jK0uDhdwlSJhxSf2DtW4XHFr6HdPaJ1BzQXQNY+mxNSjAAvW/dT4EyqI+Lw+0jfnx+9JWyzapMqD1HZR3clIcA1DypQjt7X/Ljg32sPLYh+AD1Ce+nQGl+5EfiQ/YRn4cMzMvfUZeBqzdWU/eWUSSdj/L8gCiks39PyT5vATuXrWl6nPoU30+Bpm8OBJP3MkR0TO1+qlCTL9EQlZc0RoVICq+ejUx4Auns3ZK7gD4nWrL02ZrNs0T+P2TdT4HArCMJqx9gihgQfbxUjpMQhFZpusG0Gvk/CQw3JCODY+L39mDhLgNNVxw5soCPWowZoR3kAkEgF7ASSNRzSAvmmIF7jx5adqgy99f0H1954tleiK0g/vAMgZonRyzPeokp5i9LOxESslvbUhmLrAbEMwQKGwlAkxMMc5mnYnLF52RTEccg3Ajvzf/hhU38eYiEZwikbzFmvMEQIVPWRFCZSEK2ggi4B4LGnqTpGQIda0Z+7O9pH7FHpZSuoLoaUgLewJQLMpqc5c1BNDxDoPG/twAT9jNEdNAptVfJHCQjM9FVWL4/ZMcE3hxEwzMEApMSjq5hbPa+IqeV0pj05jSevEPGQwRyTgZBkG1qiBPzvkLry/OIVwikInDiN6rHIQu4w5tRRDxLoOZDnvJjCFYTGFFH3WqaF+vd9oslHiXQs2eW/ZT1oH34X6ZCqDA+62gjL43zKIHSyXK6F8PjjxamgvrvcJ+Q73jwjQWeJFDX3QAEjmN4hDecILB6qgxKA8DPdUtxQOhQ/rpsniRQ8yTQ8ewXBSmTbcIHak232DIAgo+4MrPh97d/inV7uM+EJwkEfvngjyc2zvJNtylmKn6jhjbqSYHyQPCBoS6M/IfqrX0wi/XJHeBRAjX5sjphJgDXX7YOVs8wNxZTDo5xZeSzEPKjr4z1yR3Ak0AM/62c2kHJ1FOhw22tA6UyogpCQiOD6mTXJiZTo2Fjv+FwckZ4Eoih5OQg0EOHz4r3Bi+xHaTdRdwzDvzIcNdGAo5PbD4ssx3rkzvAfYEUpjErG9gL1P3KYp+xN0q/HtXXOjxEiekFIggdgWAm6Nt96/grONwXqN+FydY5aIJhdkdJKktDm+rLLtfL0jBs29bDVgPZmwl9GUR+KHLc9ZY1PNxirQ6H+fNwi3UTd/kTKG7G49vvgemLaBGPiLPN3fkNdlMAGho+yiCfrzJ5EGhJKjgytBb+evx12LYl3d7i9DoqC1F/JZ9sdMtTDvBTSI9YaR/GVqBJeaBrEq5NXCnSPNNjNy3i9RrTyCrUgYxmbjnKHs9pB4nKVon64LdapVwp+y6JPi1DBEvUpntsuLgfr066xnMEAl3OVJWknqy8JyZUsp208Ha6KnNDsSyTRw+R8CCBKAKTwyPVn4DW+R/SAjWHtEaFCOiqo8E7HiYQED115jfyn6/raGGnMHNvNZ8f31jgaQIBEPcLsBFok8KcgY7z4RkrEAV6OEvbMbqt63Q0uAo0prQLeYvRpz4lFmo0RoE03Gy6AaJAieOh74oYVpY5P7T/QVJRRy+kwY/1Ssx4i3msQKUiCHxKWFnmLJBoxJxhVgFdasy1WOFLjb0pK6JAt8neln8pK8tcBRLFrJ24ac+UQ4lfm5+b5ZgEkmu2Hm3kzgaiQHvGwBYb2V0xV4GmLhr+04abRdvCk+JMQbUyQxFEVGEHZs7nZpYriAK12iNTxLRhZZmrQD99H/XkWbWmtmBIxbvGILUcM/bnqyO6a7mZ5YrnVfMf/iOakYJBDaGpumwMqsaMtRhR/tjY//LmIRKIAjE/E3MKV4E6icf/eg/CtItQAY3zGWLNM8nrFp9i19hwG/QcFPjRQlaWOddiZ767pK7DpeUQXjYutmhvFghe+ZijVa6wucUuu05Cg7NA407MxQkdIamHR3Ql16jOVxPM/MDD/zRHq1xhIVCvKlaWuc/u6BZRDXENVSbr8u4SZJt6AGHujM2B7KoKt2FRBunYDaW4Mf0lPrVaf0JINk2X6wAYqjK3FGsyOVvlhgfVYoH9TeVvekKpGlZBnIB9g5+HPUBgvWWa2US+/XSB5wg07fTa+C2GSQeLC+6cgtVUyYxVi+HvAPxs1geubgBXnYEkkMk5VpZZCtQ1zhc0CV+hf+TsW6A7Ss361ecbNb4J+OnMAuWysuo+HpODQt8Dj2bvwiRxj5NfHhBXEqZFUAQBxwGZWaA4l5b4hYVAoq2sLLMUKGQxOPTk2cRRt3KpkqjZ34S52MEgjMXMAiGMzvMKokCfa1nnbpYCNTveJ21pTtRWmHiVmtvzSV03SJizEI1RrKy6D6JARZ0iHh+1hpVl1iOrUTW3ft4Esl4rmkt+66oNLDFpQtBVOsHOqtsgClQM5r0GshmjJoiL3rilf0WKNezbQR/vz3hj8Q9h70qpL8c1JQRdIRPFz7A16x6IAhW8/cKvwbcZo249NRKGBkw5Zw4IaK1nO+slx6JZ5fWHD61+2dAgnVOvYxIIRrI16x6IAoXm+R5Tf8EYVQT8oQ8A5eaAcTF68jmMQPhnLu8/3Nhf97siNytE211A18hz7d2v5m89J/IDfiHXbMM5dTVafh216iEwZF/iVLJvjBNZ9lmoLrNxFx0iCrTK8VsXXi4ZDcDvV4fYhnPvi72lKP1bcwYMI4jyCmhHxVNc7XICUaDNlSemOtiGxBHcBapODXzmAJ7YqRjelNoLdNnB4tYGAvUW8xsXVbSBlWXuAilCeyX3JxS3gnXWxbMBNLPDo+KXNefqAB3kMqj9/PO8DPsgeC37dMPQTGVJcUmABGrtBGKYimTPs3Gdfd9iN9DpAESB3kut3TmS3ex1RoFeOhV72uVD08Xa3ANivFBB1I/SKXBbgeQoJz/QnvzY3p+Vw8wgCpQwhfXrlZgEevBoC+Dz80sAPL07/vtWDn/6uUyn+aderoFKGWGbhQikZwpp1H/m4ffZ+swA8i3Gei0kk0AzqBUCPc+BfsmP+k5IcDyKLDqtU0qJGg212Nm2GCJQVs6vJmtWUTIfAyDIArFejs0kUCjVyuudCn6lFjivdbJPia8Mx87FyckbCioxG4XWuTx18Lufnl33UdrbbF1monEFCk4OBgE7h4GjVFW95Xsnvx4G4VodYdtK1H+96Gp0vvfZt0OOLHtl0rqVT7J12h5kgRaztcxYSD+fmpQ5E4Al1Iqng073pvwa4jihsa/loRp/2sWZ4zuQWfDkh7/1fzZxNFuv7bg/TxQDEjd+mOhiFkIctdeC/imHdRa6WuVq6Xwm9VHyN9klCWQ7298eRIF6pShUKQw7IzjBaUNRNGi8yw3/8iFBaKj+qnVFhl9wtcdSMtXWOq0/vfsbWSAKdHVpUMulF1lZdntZeGsy+2AEoYbWEHWu9kd8IemZ7uFfnSZr+tbH3PQBWSBqn03abqQocBLovdI6ZbppKVN3qv4iimzr+TLmxy40nlq1MQRMTho36dQgDj5YgyjQyo8feOCTLawscxFocPGhTedxXDzW+L0O4kpbfbDbaWjGei2ch75vp0NYTX+BbOZxcxFoR9refddK5Tc1IcYAamxMbi3QSchujgAKU9KPpTi4NI8ZF9MTnbnk2qUl5H2FKy89TwX0JOyqenX5W3y7+nR0IAg+xrCjAWh8geiLdQbvse2QvVRxsCAvO0mlhnUYUUf9nw4kxbLpr17g/ZFiGNWyms88da1xBRqWeejkj/MWGNoL/ZLsO2RhGi2RX5mlUn9LFdA60rloiGutFfqWd1fDqc0r5zNv/tWoArU5HgSeLV47PvE16pupQ9Z2TcJOc5/Ad7FMK1NqTxp7YEsBwG1GDmFWJ75dfTHCBwTGPcwY16gCvfEx2YrrdQ0E6KeJHV19LmXolu/8jo0LjK3OiDGvU343R117WEdc1a8/qO08xlYgmctqnjULz0Rnv8YchSRQcORdTFf4s9vTgCd+CkBG1xwAqHsHFOVPXFKXPWFXtGjD64O2DMlNPhJhLCfXa8naS6aT6IXJtX0iJONtVwUaDnd9RRIo7cdu/gGP/pddx4ZBoFap7UFaxFggooZu+pclblp4RluZK1PWftMRvi/p8LtYWbNWXwL7R+ggriIM+yRjNr0xeS+n5xV160N19wN42r0DSaBKQ71RwMoyUyE9KCUxL29Q313Uo74FWaKBr86GoriTe++tLrpWVx95bnTacwqVWp1M9SUu6MXADR0NK4HSnZ42KG37upxBD8bFJ//By/JWJIHSyRzk/8h/EVuwRhy1gwatWjeC+rdHJfmRWt8yflE2rFLsFlfVFc/ITyNreDLbxJNeBRRCagNgaIfzTRW2DSZFOp277WXwymZW/joArQzaWoRhdyODWVl22VA8WrBoizrnoWjwecVteK3wHZ04TE7gMP4JXAcxOVmZ79dP6tARNu2g2U6tZpJ/O6iuD1kTDqSs/HWA+7XYgMuaqGYMDxZct6TnxO0aKP4i9W3xF79eGdjiIlRgxfSbST6yMzUXGCcU1rUY4XiUlySNLA42/J0MwO+9xObA5l9veZ3TxfEh0F/vNN+4lS5Qh9F64mNRft3zL7HqxP7qleNB6i8XCB1hnJhokEQ3KJU8tH3eAf9yZnDu2gBweU/0ELD9zzBT2MO1qeHFHJ+duS9QGTURozdNoOfW6Pk7EdVCp3bDz5dlnVt+pkKaZckn+s/z421nmFHPz5xam5947MLwdjFJpbGG9WYBkccl1zoB/6KxTn/mCPcFKuoOwPgT9luLsdtPekhkctG0GnmFzjKf1TCzrIiSxHpgw+XS5wdPvTlo5c/GLyvfBmUT4gEQX2DhjwUkga4bPWOMfE+yEYDNF9zfHqfjj+cxagE4Fm5zSykzi+lfNa7nkgd/9P1rVNNk2JwRokwAbr556WFwdQZLfwwgCeSTNt5JbI9RZA4Yudw2mMvjjsdSldB+qBBqZPQg5LmkftmX/tgSm0m24MXFL+67y20tJ9otNm8Ee8scn0mLovTlTIXdmLwlQ6GaOpn06LhTy+Kmkn0tSf7+IE7ueNgDMwPdtuA2IxnWI/SIb4N47vRoAJ7/o3zb8YQj3B8AeKJAAPheoPYFVNAUoumDo72bBczY/gEAraGbK/BYPZOuPcjif8LNYZ/H1FQjkekWq0Ac7BqwM3VS+xV33HxLAKJAi9e1Dlq3sMveBHTLbo+LRZAtxXIGgXTVzn/n8/K80c9dvHP19fU7dp4rZH4MxsILxI0FfMl8nw/aIs1eMuD+ftI9FFqdQm57h0G4w+mvfA+tePUXzQ9PfF767H8+fZPlxEp7EAUq9yNrzQrQBrkG4WXDbb9cXK1Q2ApEnHP6o+lkoXMTqxKv2ZzltgMAWaDoDUFB66M7RrGY9sfLjuSZ0qun7e8xpz8JexrMVO376/HSPWhvInMB8s4LUtX+Nq9vdjxtzg5eBBLVYkV2U4Gdl0ELpoNzR4rfBTuUdm1XLnhmNW+hhab+pm13vthx8kcWfvpM+iJ5ISY7LGdRHDgBUaDgGJk8pnE2WLLhQ51GYiOQ4/1KnzsV8lLc7Pwcce9tZ+rs5v5zAlGg2LDWrTewW77D11sRirGKfyBhVU47mms/ShJ76kvfixvBHvEd4kt+To86/YWs5v0aZ/8gWwYRBbazFB2kbH8sC4jWvkbci8sY0Y2v90t4vkBARhTmWAmkdpBwyvwTgWBwWVly36Dk1XN5OjvyLRYUFNZIe5jZMgZKbLocDhJO+mhUVt4J7GL32GO3j/I1wwG1kD4gV8TyPKqBDK7TIAnUKqVzzrrc88t+AWDFJ3yd3NOreYqlNr1WzFHCp08VZXx3Ixce75bM2xYoHrPzgjMIaNUUctwh7HcrWVZdjF0s5aeKp/CGHAQkeBJdIInDhFGVNwmJovYfHl9j17iLWTgywaa7yvSCJIqm0gKcwLAZkkreTt3IazW4ooAqukAjmdKEphy7SQ0yqqBWy+NLtrxDoKPlViNjTBNSx+wIXKmCZVCDwQrI416LjbyYhSODcRX9HmOa+rO/ffvDhERK6qMp5fNNml5RSAOgsRp/ZkqR8MDQVYRGR+Y0Rb2YKQFHvESgMnm9C4E+WNin1hB7R8dYRnHEfYHsxqX7ztOTEe+uZRofWA2UMU0dE0WUSQn9Fha6pTyemAeB7Malu72pJ/mwu5bp4PRbjGlOfEDoXWWYBsc0WDW7/R5dwMMt5mBcmt/XqdMGERlvMb+kj3OkxdjcY+XZWl63GPKSMghcgLStAhl6ze/MB2OvKwidAl8FX+XzxN4i0DB6W5rhRZlrnwFgJwEJrFjh6HERN7xFIEC/xRjaQTPfB98q6+sw3ZV6dpNxXeE1AtFrsVD76IDkpViJLvsAXh/H7w5MXiNQIU0gpulcza4QS9PujiKOFTpcVcAJrxFogIuuxvMr5RHzP6gm0vnsZwAvEgjQCmmmidIhi/fpZGfglYG2b0h0E+8RiPbUdSZDdHB2RrxcJ55w0ul+DuzxHoGqLBU943TM3LPni4tV6Tzr40UCTbLkoEeY4k9Oi+gMKg4P5vesXiQQrSV0iyF2SJXiVkLfzNks5sAh4UUCWQohBsuvVtWoJbeln8ziY9cpOl4k0CWzQAxTnktTwutwvKIkk99WkFcJNMZcSNs/DwqoKZFiGqhRruf5pN4kEDCvcym3iwqsxOqvbYKaf5xPz+OANwlknovHMKpTYNgPt/QK3yf1JoEUTsqg86Q+OhkkeJu0YMKbBDprEoghTqd/SYAWuj0v2hZvEijQsUCPQznV0Nbx3FMF3iWQualov/z0EK7B885CgpeFzlZ4lUCm8Wf7gcE4FYFRY4u7eD+nVwl0xSiQ/TZ4U1TwXikhhq62WGSPVwn0gkEfhhlmoksEQeDl/BdB3iWQsRBimga8XapRS1UH+T+lNwrElE+kOsm9arXzXSs44b5AjjYXagiBCEcCPQJ/25anJPgvgngQyNHmQg0h0D1H1fwO7XIfEAYb4D3H7gtkt7nQi9v0XEHemoIFjkrpg9LItITrDVBG8yCQ3eZCQYP0fMt6ThoCBoHsp/y/W3/p+Gl5Q7wYiYcyyMHmQtMb4qXMhuWH9h0uv7/Ed2vuubnAmZGGq8UaRCBwiIBKuxdRkQQu3P0lx70VnONtAjU6gkAuaDiBPoWpDiiV1LhGiZAGKZFEjpCo/pojZxXO95RrECJQ3piC9E5nlET9IxASTZ+OcrrGQhDIBYJALhAEcoEgkAsEgVwQ3hshUQqKJZREvVGmnE2dinK6xgJp7KpxE/k5WsQoICAgICAgICAg4BG0S69KN41/9FO4TDT7btl5xo6bJQndIns7iB41HpGrRKsiDYet7Dcxt03UXdIDzLzk3A7NIgc7iB41HkXdQLe7+iPR4cmO3DEnGrURgOYa53YsR1zsIHrUeGD+wN8wlfnLMIf7pVgSkb3IDdHOk9ATs7eD6FGjsBjCaZQ7AXp3RmT6M7ljnQiAV/O2Mb6hx5KElthxIod2gEuPGpmiR8CjhdTBSv0YMvOLF82JRJvSHDw6MiehHXGxg+hR4xG5DFCb1BleSOewkDYlGnHB0WMaix3jEUc7iB41Hu0yCtLamhxx5I450XJH88lodoxHHO0geiQgICAgICAgICAgICAgICAgICAgICAg8H9LgLgA9fVN+2wGG6DDGOtE5J+DhX7kH44u3l8mXROjDtFBhwKRRww7S9OioC9g3PfVCzi6JHIHAAnvAbB5Vauo8oJF5suGayq3/HJdvAy03FVQcqQriIY5fvCbK1Vfge2fAfDbIuraQ2+X3P3MEAOCY0pLY4LJ371/qUr/1r62MVUlP+pVJRPEwJw2BjsAzq9raTrUp213pOTOTNNpPIyHtJ1elDYHU1KBX+Vjuzf6+YZZBHqhEwwFvbVg91iRKCTLkIM+Br204OVToEn1Q9T3vFDQWWvMWwfW+/mujyGPPgR9tJSB/ZEB/puhKedBi53vWpgP9WljvhF1lwaZwjyLpanAtzwUBNa2H3cCSIOol3qboqAP9Ye8MBk1GmqUIZD6J6C6Q0iq/rvvsLlbTNcvp34tNaUhUbQhcxFNILOdFpZDfVpZKwBa+prCPArRDcqrEwDsnB81DSjIS2xlEchwnRAoWpBKdKQFgN1zo2bqD47smNjddP3Ur1srzGkAkJMCtaEJRLNDOyT/KslvPZqZwjyK51XtABhMdAUjzxc1AX+G+fquhbT5Awb/D2zw91kebchShvCQzKqW+gNVMBgJRYaY2J98DbcYMAp0ICLAL5wmkNkOoB+Sf48sFT1aG2wK8yh26JdwZS0DvqUbyP//PyvyPzLXVWaB2u4rK4ltC0Buvp/x4pvK9hkOltzOXntpmyEmOJYspNvSBGoXJy5aTBPIZIf8Qjsk/3ZKLS0MNYd5MBPJDN65FiXlOabN7P/9hP8R1GbPIdfp/PpUBTS8Nx5Iq/1K5dFOrtO9qWyADaYEBAQEBAQEBAQEBAQEBAQEBATY8D8HPX5RNQN5NQAAAABJRU5ErkJggg==" alt="plot of chunk unnamed-chunk-4"/> </p>

<p>The <code>colon_curves</code> object is another <code>GRanges</code> object which contains the methylation data used to define these regions.</p>

<pre><code class="r">show(colon_curves)
</code></pre>

<pre><code>## GRanges with 114949 ranges and 7 metadata columns:
##            seqnames               ranges strand   |        id    type
##               &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt;   | &lt;numeric&gt; &lt;array&gt;
##        [1]     chr1       [15865, 15865]      *   |         1 OpenSea
##        [2]     chr1       [18827, 18827]      *   |         2 OpenSea
##        [3]     chr1       [68849, 68849]      *   |         4 OpenSea
##        [4]     chr1       [69591, 69591]      *   |         5 OpenSea
##        [5]     chr1       [91550, 91550]      *   |         6 OpenSea
##        ...      ...                  ...    ... ...       ...     ...
##   [114945]     chrY [25114810, 25114810]      *   |    224593 OpenSea
##   [114946]     chrY [25314171, 25314171]      *   |    224594 OpenSea
##   [114947]     chrY [26716289, 26716289]      *   |    224595 OpenSea
##   [114948]     chrY [27009430, 27009430]      *   |    224596 OpenSea
##   [114949]     chrY [27210334, 27210334]      *   |    224597 OpenSea
##            blockgroup      diff    smooth normalMean cancerMean
##             &lt;numeric&gt;  &lt;matrix&gt; &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt;
##        [1]          1 -0.010675      &lt;NA&gt;    0.87121    0.86054
##        [2]          1 -0.003764      &lt;NA&gt;    0.64226    0.63849
##        [3]          1 -0.065629      &lt;NA&gt;    0.54548    0.47985
##        [4]          1 -0.227616      &lt;NA&gt;    0.72965    0.50203
##        [5]          1 -0.022469      &lt;NA&gt;    0.05758    0.03511
##        ...        ...       ...       ...        ...        ...
##   [114945]       1049 -0.014063      &lt;NA&gt;     0.5861     0.5720
##   [114946]       1049 -0.073670      &lt;NA&gt;     0.3947     0.3210
##   [114947]       1050 -0.001695      &lt;NA&gt;     0.1776     0.1759
##   [114948]       1051 -0.028182      &lt;NA&gt;     0.4106     0.3825
##   [114949]       1051  0.045758      &lt;NA&gt;     0.3767     0.4225
##   ---
##   seqlengths:
##     chr1  chr2  chr3  chr4  chr5  chr6 ... chr20 chr21 chr22  chrX  chrY
##       NA    NA    NA    NA    NA    NA ...    NA    NA    NA    NA    NA
</code></pre>

<p>This basepair resolution data includes a mean methylation levels for normal and cancer and a smoothed estimate of methylation difference. This smoothed difference estimate is used to define regions in the <code>colon_blocks</code> object.</p>

<p>Finally, the <code>apColonData</code> object is an <code>ExpressionSet</code> containing gene expression data for colon normal and tumor samples for genes within regions of methylation loss identified <a href="http://www.nature.com/ng/journal/v43/n8/full/ng.865.html">this paper</a>. Our goal in this vignette is to visualize this data as we browse the epigenome.</p>

<h3>The epivizr session manager</h3>

<p>The connection to <code>Epiviz</code> is managed through a session manager object of class <code>EpivizDevMgr</code>. We can create this object and open <code>Epiviz</code> using the <code>startEpiviz</code> function.</p>

<pre><code class="r">mgr = startEpiviz(workspace = &quot;C60FA3168F34DBC763F579C1EADA8AF0&quot;)
</code></pre>

<p>This opens a websocket connection between the interactive <code>R</code> session and the browser client. This will allow us to visualize data stored in the <code>Epiviz</code> server along with data in the interactive <code>R</code> session.</p>

<h3>Adding block region devices</h3>

<p>Once the browser is open we can visualize the <code>colon_blocks</code> object containing blocks of methylation modifications in colon cancer.
We use the <code>addDevice</code> method to do so. We then use the <code>service</code> function to let the interactive <code>R</code> session serve data requests from the browser client. We need to escape (using <code>ctl-c</code> or <code>esc</code> depending on your platform) to continue with the interactive <code>R</code> session.</p>

<pre><code class="r">blocks_dev &lt;- mgr$addDevice(colon_blocks, &quot;450k colon_blocks&quot;)
mgr$service()
</code></pre>

<p>You should now see that a new track is added to your <code>Epiviz</code> session. You can think of this track as an interactive display device in <code>R</code>. As you navigate on the browser, data is requested from the <code>R</code> session through the websocket connection. Remember to escape to continue with your interactive <code>R</code> session. The <code>blocks_dev</code> object inherits from class <code>EpivizDevice</code>, which serves as the fundamental type for display devices in <code>epivizr</code>. Note that the &ldquo;brushing&rdquo; effect we implement in <code>Epiviz</code> works for <code>epivizr</code> tracks as well.</p>

<p>The point of having interactive data connections to <code>Epiviz</code> is that we may want to iteratively compute and visualize our data. For example, we may want to only display methylation blocks that at a certain statistical significance level.</p>

<pre><code class="r"># subset to those with pvalue &lt;0.05
keep &lt;- colon_blocks$p.value &lt; 0.05
mgr$updateDevice(blocks_dev, colon_blocks[keep, ])
</code></pre>

<p>Now, only this subset of blocks will be displayed in the already existing track.</p>

<h3>Adding line plots along the genome</h3>

<p>We have a number of available device types in <code>epivizr</code>, which are listed using the <code>listTypes</code> method. In the previous section, we used the <code>block</code> type. For the methylation data at base-pair resolution in the <code>colon_curves</code> object, we will use the <code>bp</code> type.</p>

<pre><code class="r"># add low-filter smoothed methylation estimates
means_dev &lt;- mgr$addDevice(colon_curves, &quot;450kMeth&quot;, type = &quot;bp&quot;, mdCols = c(&quot;cancerMean&quot;, 
    &quot;normalMean&quot;))
mgr$service()
</code></pre>

<p>Notice that we added two lines in this plot, one for mean methylation in cancer and another for mean methylation in normal. The <code>mdCols</code> argument specifies which columns in <code>mcols(colon_curves)</code> will be displayed. </p>

<p>We can also add a track containing the smooth methylation difference estimate used to define methylation blocks.</p>

<pre><code class="r">diff_dev &lt;- mgr$addDevice(colon_curves, &quot;450kMethDiff&quot;, type = &quot;bp&quot;, mdCols = c(&quot;smooth&quot;), 
    ylim = matrix(c(-0.5, 0.5), nc = 1))
mgr$service()
</code></pre>

<p>We pass limits for the y axis in this case. To see other arguments supported, you can use the help framework in R <code>?epivizr::addDevice</code>.</p>

<h3>The session manager again</h3>

<p>We can use the session manager to list tracks we have added so far, or to remove devices.</p>

<pre><code class="r">mgr$listDevices()
mgr$rmDevice(diff_dev)
</code></pre>

<h3>Adding a scatterplot</h3>

<p>Now we want to visualize the colon expression data in <code>apColonData</code> object as an MA plot in <code>Epiviz</code>. To do so we will add a device of type <code>gene</code> which expects an <code>ExpressionSet</code> object as input. So first, we construct an <code>ExpressionSet</code> with <code>M</code> and <code>A</code> columns.</p>

<pre><code class="r">keep &lt;- pData(apColonData)$SubType != &quot;adenoma&quot;
apColonData &lt;- apColonData[keep, ]
status &lt;- pData(apColonData)$Status
Indexes &lt;- split(seq(along = status), status)

exprMat &lt;- exprs(apColonData)
mns &lt;- sapply(Indexes, function(ind) rowMeans(exprMat[, ind]))
mat &lt;- cbind(colonM = mns[, &quot;1&quot;] - mns[, &quot;0&quot;], colonA = 0.5 * (mns[, &quot;1&quot;] + 
    mns[, &quot;0&quot;]))

eset &lt;- ExpressionSet(assayData = mat, annotation = annotation(apColonData))
show(eset)
</code></pre>

<pre><code>## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 4154 features, 2 samples 
##   element names: exprs 
## protocolData: none
## phenoData: none
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
## Annotation: hgu133plus2
</code></pre>

<p>Including the annotation type in this object is essential since <code>epivizr</code> will use that annotation to determine genomic locations for this data so that only probesets inside the current window are displayed. </p>

<pre><code class="r">eset_dev &lt;- mgr$addDevice(eset, &quot;MAPlot&quot;, type = &quot;gene&quot;, x = &quot;colonA&quot;, y = &quot;colonM&quot;)
mgr$service()
</code></pre>

<p>In this case, we specify which data is displayed in axis of the scatter plot using arguments <code>x</code> and <code>y</code>.</p>

<h3>The SummarizedExperiment Object</h3>

<p><code>Epiviz</code> is also able to display plots of data in the form of a <code>SummarizedExperiment</code> object. If we load the <code>tcga_colon_expression</code> dataset, we can see that the <code>colonSE</code> object contains information on 239322 exons in 40 samples.</p>

<pre><code class="r">data(tcga_colon_expression)
show(colonSE)
</code></pre>

<pre><code>## class: SummarizedExperiment 
## dim: 239322 40 
## exptData(0):
## assays(1): &#39;&#39;
## rownames(239322): unknown1 141710 ... 273933 273934
## rowData metadata column names(1): exon_id
## colnames(40): TCGA-A6-5659-01A-01R-1653-07
##   TCGA-A6-5659-11A-01R-1653-07 ... TCGA-D5-5540-01A-01R-1653-07
##   TCGA-D5-5541-01A-01R-1653-07
## colData names(110): bcr_patient_barcode bcr_sample_uuid ...
##   Basename fullID
</code></pre>

<p>The <code>assay</code> slot holds a matrix of raw sequencing counts, so before we can plot a scatterplot showing expression, we must first normalize the count data. We use the geometric mean of each row as a reference sample to divide each column (sample) by, then use the median of each column as a scaling factor to divide each row (exon) by.</p>

<pre><code class="r">ref_sample &lt;- 2^rowMeans(log2(assay(colonSE) + 1))
scaled &lt;- (assay(colonSE) + 1)/ref_sample
scaleFactor &lt;- matrixStats::colMedians(scaled)
assay_normalized &lt;- sweep(assay(colonSE), 2, scaleFactor, &quot;/&quot;)
assay(colonSE) &lt;- assay_normalized
</code></pre>

<p>Now, using the expression data in the <code>assay</code> slot and the sample data in the <code>colData</code> slot, we can plot mean exon expression by sample type.</p>

<pre><code class="r">status &lt;- colData(colonSE)$sample_type
index &lt;- split(seq(along = status), status)
logCounts &lt;- log2(assay(colonSE) + 1)
means &lt;- sapply(index, function(ind) rowMeans(logCounts[, ind]))
mat &lt;- cbind(cancer = means[, &quot;Primary Tumor&quot;], normal = means[, &quot;Solid Tissue Normal&quot;])
</code></pre>

<p>We just create a new <code>SummarizedExperiment</code> object with the two column matrix. All the information about the features of interest, in this case exons, are stored in the <code>rowData</code> slot and are automatically queried by <code>Epiviz</code>.</p>

<pre><code class="r">sumexp &lt;- SummarizedExperiment(mat, rowData = rowData(colonSE))
se_dev &lt;- mgr$addDevice(sumexp, &quot;Mean by Sample Type&quot;, type = &quot;gene&quot;, x = &quot;cancer&quot;, 
    y = &quot;normal&quot;)
mgr$service()
</code></pre>

<p>Again, the x and y arguments specify what data will be displayed along which axis.</p>

<h3>Closing the session</h3>

<p>To close the connection to <code>Epiviz</code> and remove all tracks added during the interactive session, we use the <code>stopServer</code> function.</p>

<pre><code class="r">mgr$stopServer()
</code></pre>

<h3>SessionInfo</h3>

<pre><code class="r">sessionInfo()
</code></pre>

<pre><code>## R version 3.0.1 Patched (2013-07-02 r63147)
## Platform: x86_64-unknown-linux-gnu (64-bit)
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats     graphics  grDevices datasets  utils     methods  
## [8] base     
## 
## other attached packages:
##  [1] antiProfilesData_0.99.1 Biobase_2.21.6         
##  [3] devtools_1.3            epivizr_0.99           
##  [5] knitr_1.2               rjson_0.2.12           
##  [7] GenomicRanges_1.13.26   XVector_0.1.0          
##  [9] IRanges_1.19.15         BiocGenerics_0.7.2     
## [11] httpuv_1.0.6.3          Rcpp_0.10.4            
## 
## loaded via a namespace (and not attached):
##  [1] digest_0.6.3      evaluate_0.4.4    formatR_0.8      
##  [4] httr_0.2          markdown_0.6.1    matrixStats_0.8.1
##  [7] memoise_0.1       R.methodsS3_1.4.4 RCurl_1.95-4.1   
## [10] stats4_3.0.1      stringr_0.6.2     tools_3.0.1      
## [13] whisker_0.3-2
</code></pre>

</body>

</html>
